File: exec/src/mcp/ulid.ts
/**
 * Crockford Base32 ULID，无外部依赖。移植自 `sandbox/mcp/ulid.py`。
 *
 * 与正式域 ID 兼容：26 个大写 Crockford 字符。不是 UUID，也永远不会产生
 * `exec_...` 前缀。同一毫秒内单调递增——靠对熵值加一，而不是重新取随机数，
 * 这样同毫秒内产生的多个 ID 仍然保持可排序。
 *
 * ## Model Experience
 * 模型看不到这个模块。它产出的 ID 会作为 `context_id` / `artifact_id` 出现在
 * 工具结果里，26 字符固定长度，对 token 与 KV cache 的影响是常量。
 *
 * ## Known Limitations and Deferred Work
 * - 单进程内单调。多进程下不同进程可能在同一毫秒产出不同熵值的 ID，这不影响
 *   唯一性（80 位熵），只影响跨进程的严格排序。
 */

const CROCKFORD = '0123456789ABCDEFGHJKMNPQRSTVWXYZ';

/** 熵为 80 位（10 字节）；时间部分 48 位（毫秒）。 */
const RANDOM_BYTES = 10;

let lastMs = -1;
let lastRandom: Uint8Array | null = null;

function encodeTime(ms: number): string {
  if (ms < 0 || ms >= 2 ** 48) throw new RangeError('ULID time component out of range');
  const chars = new Array<string>(10);
  let rest = BigInt(ms);
  for (let i = 9; i >= 0; i -= 1) {
    chars[i] = CROCKFORD[Number(rest & 0x1fn)] as string;
    rest >>= 5n;
  }
  return chars.join('');
}

function encodeRandom(raw: Uint8Array): string {
  if (raw.length !== RANDOM_BYTES) throw new RangeError('ULID entropy must be 10 bytes');
  let n = 0n;
  for (const byte of raw) n = (n << 8n) | BigInt(byte);
  const chars = new Array<string>(16);
  for (let i = 15; i >= 0; i -= 1) {
    chars[i] = CROCKFORD[Number(n & 0x1fn)] as string;
    n >>= 5n;
  }
  return chars.join('');
}

/** 就地对 10 字节熵加一；溢出返回 false。 */
function incrementRandom(buf: Uint8Array): boolean {
  for (let i = buf.length - 1; i >= 0; i -= 1) {
    const byte = buf[i] as number;
    if (byte === 0xff) {
      buf[i] = 0;
      continue;
    }
    buf[i] = (byte + 1) & 0xff;
    return true;
  }
  return false;
}

function randomBytes(size: number): Uint8Array {
  const out = new Uint8Array(size);
  globalThis.crypto.getRandomValues(out);
  return out;
}

/** 生成一个新的 ULID（同毫秒内单调，大写 Crockford）。 */
export function newUlid(): string {
  let ms = Date.now();
  if (ms > lastMs) {
    lastMs = ms;
    lastRandom = randomBytes(RANDOM_BYTES);
  } else {
    // 同一毫秒，或时钟回拨 → 熵值加一。
    if (lastRandom === null) lastRandom = randomBytes(RANDOM_BYTES);
    else if (!incrementRandom(lastRandom)) {
      throw new Error('ULID entropy exhausted within the same millisecond');
    }
    ms = lastMs;
  }
  return encodeTime(ms) + encodeRandom(lastRandom);
}

File: sandbox/mcp/ulid.py
"""Pure-Python Crockford Base32 ULID generator (no external deps).

Compatible with formal domain IDs: 26 uppercase Crockford characters.
Not UUID and never produces ``exec_...`` prefixes.
"""

from __future__ import annotations

import os
import threading
import time

_CROCKFORD = "0123456789ABCDEFGHJKMNPQRSTVWXYZ"
_CROCKFORD_LEN = 32

# Entropy is 80 bits (10 bytes); time component is 48 bits (ms).
_RANDOM_BYTES = 10

_lock = threading.Lock()
_last_ms: int = -1
_last_random: bytearray | None = None


def _encode_time(ms: int) -> str:
    if ms < 0 or ms >= (1 << 48):
        raise ValueError("ULID time component out of range")
    chars = ["\0"] * 10
    for i in range(9, -1, -1):
        chars[i] = _CROCKFORD[ms & 0x1F]
        ms >>= 5
    return "".join(chars)


def _encode_random(raw: bytes) -> str:
    if len(raw) != _RANDOM_BYTES:
        raise ValueError("ULID entropy must be 10 bytes")
    # 80 bits → 16 Crockford chars
    n = int.from_bytes(raw, "big")
    chars = ["\0"] * 16
    for i in range(15, -1, -1):
        chars[i] = _CROCKFORD[n & 0x1F]
        n >>= 5
    return "".join(chars)


def _increment_random(buf: bytearray) -> bool:
    """Increment 10-byte entropy in place; return False on overflow."""
    for i in range(len(buf) - 1, -1, -1):
        if buf[i] == 0xFF:
            buf[i] = 0
            continue
        buf[i] = (buf[i] + 1) & 0xFF
        return True
    return False


def new_ulid() -> str:
    """Generate a new monotonic-within-ms ULID (uppercase Crockford)."""
    global _last_ms, _last_random
    with _lock:
        ms = int(time.time() * 1000)
        if ms > _last_ms:
            _last_ms = ms
            _last_random = bytearray(os.urandom(_RANDOM_BYTES))
        else:
            # Same ms or clock went backwards → bump entropy.
            if _last_random is None:
                _last_random = bytearray(os.urandom(_RANDOM_BYTES))
            elif not _increment_random(_last_random):
                raise RuntimeError("ULID entropy exhausted within the same millisecond")
            ms = _last_ms
        return _encode_time(ms) + _encode_random(bytes(_last_random))


__all__ = ["new_ulid"]
